compiler/rustc_lint/src/array_into_iter.rs

```diff
@@ -37,10 +38,8 @@ declare_lint! {
     "detects calling `into_iter` on arrays",
     @future_incompatible = FutureIncompatibleInfo {
         reference: "issue #66145 <https://github.com/rust-lang/rust/issues/66145>",
-        edition: None,
-        future_breakage: Some(FutureBreakage {
-            date: None
-        })
+        reason: FutureIncompatibilityReason::EditionSemanticsChange(Edition::Edition2021),
+        custom_explanation: Some("This will continue to compile in Rust 2021 but it will behave slightly differently!")
```

Comment:
> This will continue to compile in Rust 2021 

In many cases, it will result in the compilation failing because of a type mismatch. The `array.into_iter()` expression might still compile, but the code it appears it probably won't.

Reply:
Removing `future_breakage` here is going to cause problems for Cargo's testsuite, which uses this for testing the future-incompat system.  Can you keep it around?

And, maybe I'm out of the loop, but I was under the impression this lint was going to change to be a hard error on all editions?

Hm, I also checked, and this doesn't migrate with `cargo fix`. 😦   Is that intended to work?

And, maybe not for this PR, but it might be nice to fold `future_breakage` into `FutureIncompatibilityReason`, unless there is some reason to keep them independent?

Reply:
FWIW, I'm going to try looking into making the future-reporting testing easier with a custom -Z flag or something. I didn't want it to sound like this is important to keep as a future-reportable lint. Just wanted to make it known that it will affect Cargo.  Worst case I can temporarily disable all of those tests. I'm pretty sure the tests on rust-lang/rust will pass since the tests should be ignored here.

Reply:
> Can you keep it around?

Yes, I'll add it back.

> I was under the impression this lint was going to change to be a hard error on all editions?

This was not my understanding as this is simply a papercut (not a bugfix), and breaking existing code in Rust 2015/2018 would not be acceptable. @m-ou-se can you confirm?

> it might be nice to fold `future_breakage` into `FutureIncompatibilityReason`

I agree, but I might leave that for a future PR.

Reply:
> Removing `future_breakage` here is going to cause problems for Cargo's testsuite, which uses this for testing the future-incompat system. Can you keep it around?
> 
> And, maybe I'm out of the loop, but I was under the impression this lint was going to change to be a hard error on all editions?

We are doing this now over an edition to not cause future breakage. We *might* want to consider removing the hack from the older editions at some point, but I don't see that happen any time soon. (Otherwise we didn't need to tie it to the edition.)

So yeah, I'd say this is no longer a `future_breakage`. We don't need to warn about dependencies breaking in a future release, because they will just stay on the same edition.

> Hm, I also checked, and this doesn't migrate with `cargo fix`. frowning Is that intended to work?

Oh, it should suggest to replace `array.into_iter()` by `array.iter()`.

Reply:
Oh, sorry, I think I had this lint confused with another one in terms of how much breakage it causes.

And I rechecked `fix`, and it is working, I'm not sure why I was confused yesterday.  So you should ignore almost everything I've said here. 🤣  I do plan to pursue a better testing strategy for future-reporting so it doesn't need to use this lint.

Thanks, and apologies for the confusion!  (And FWIW I think this PR is great and helps clarify both the code and the messages.)

Reply:
Posted #86478 to decouple Cargo's tests from array_into_iter.